controllers/mongodb_tls_test.go

```diff
@@ -328,7 +328,69 @@ func TestPemSupport(t *testing.T) {
 		err = r.ensureTLSResources(mdb)
 		assert.Error(t, err)
 		assert.Contains(t, err.Error(), `if all of "tls.crt", "tls.key" and "tls.pem" are present in the secret, the entry for "tls.pem" must be equal to the concatenation of "tls.crt" with "tls.key"`)
+	})
+}
+
+func TestTLSConfig_ReferencesToCACertAreValidated(t *testing.T) {
+	t.Run("Success if reference to CA cert provided via secret", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
+			Name: "certificateKeySecret"}, nil)
+
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Success if reference to CA cert provided via config map", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, &mdbv1.LocalObjectReference{
+			Name: "caConfigMap"})
+
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
 
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Succes if reference to CA cert provided both via secret and configMap", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
+			Name: "certificateKeySecret"}, &mdbv1.LocalObjectReference{
+			Name: "caConfigMap"})
+		mgr := kubeClient.NewManager(&mdb)
+		cli := mdbClient.NewClient(mgr.GetClient())
+		err := createTLSSecret(cli, mdb, "CERT", "KEY", "")
+
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Failure if reference to CA cert is missing", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, nil)
+
```

Comment:
There's quite a lot of code duplication within this test. Can I ask you to re-implement it to the parameters-driven test, like https://github.com/mongodb/mongodb-kubernetes-operator/blob/master/test/e2e/replica_set_arbiter/replica_set_arbiter_test.go#L28 ?

Reply:
Yes sure. I initially started it as a table drive test, but skimming through the codebase it didn't seem to me like a prevalent pattern!